### DISCUSSION TITLE
Is there any useful documentation about how to write filters?

### DISCUSSION THREAD

**Discussion by raquelhortab** [fail2ban/fail2ban#3842] [Q&A] [answered]
Considering how widely used this tool is, I find it strange that I cannot find any good documentation about how to write filters.

For instance, I want to detect this kind of requests:

```
[a991b1] Started GET "/services/getFile.cmd?userfile=config.xml" for xx.xxx.xx.xx at 2024-09-25 09:53:50 +0200
[a991b1] Processing by PagesController#show as 
[a991b1]   Parameters: {"userfile"=>"config.xml", "url"=>"services/getFile"}
[a991b1] PC not signed in. Scope: , token 
[a991b1] Completed 406 Not Acceptable in 0.4ms (ActiveRecord: 0.0ms)
```

I managed to make a multiline filter work by detecting the request id (so it groups lines by request). I managed to do that by looking at some examples and blogs (unrelated to fail2ban).

```
prefregex = ^<F-MLFID>\[\w{6}\] </F-MLFID><F-CONTENT>.+</F-CONTENT>$

failregex = ^<F-NOFAIL>Started .+ for <HOST> at .*$</F-NOFAIL>$
            ^Completed 406 Not Acceptable
```

However, I need to only match those that contain both "PC not signed in" and "Completed 406 Not Acceptable".

I have tried a bunch of things like this, along with `^` or `\s`

`PC not signed in.*\n.*Completed 406 Not Acceptable
`

I also tried them on two separate lines but then when any of the two is matched, the filter is triggered:
```
^PC not signed in.
^Completed 406 Not Acceptable
```

I have not found any documentation that gives any directions whatsoever on how to write filters and I think that would be highly useful, I am wasting lots of time trying to figure this (and other configs) out

**Comment by raquelhortab** [comment]
I am thinking that ignoreregex could work if I set it to "PC signed in" but still I think there must be another way and some explanation somewhere

**Reply by raquelhortab** [reply]
no, ignoreregex does not work, I guess it does not do what I thought

**Comment by sebres** [comment, marked as answer]
Every single RE match will cause a finding, and after `maxretry` attempts - a ban.
One could either increase `maxlines` and match several lines (bypassin in-between lines with `<SKIPLINES>` tag)… so both lines would be matched with single RE.
Or one could capture "PC not signed" with `...<F-X>PC not signed</F-X>...` in the RE matching this (it becomes then stored in the ticket)… And then check with `ignorecommand` whether tag `<F-X>` was set.
